src/server/server.rs

```diff
@@ -173,27 +144,28 @@ impl<T: Send + Sync + 'static> Server<T> {
     /// process supervisor of an individual node and the replacement and
     /// decommissioning logic of the cluster.
     ///
-    #[cfg(not(feature = "rustls"))]
-    pub fn listen<A: ToSocketAddrs>(
-        self,
-        address: A,
-    ) -> impl Future<Output = Result<ExitCode, DynError>> {
-        // Create a HttpAcceptor to serve connections over HTTP.
-        let acceptor = acceptor::http::HttpAcceptor;
+    pub async fn listen(&self, address: impl ToSocketAddrs) -> Result<ExitCode, DynError> {
+        let listener = TcpListener::bind(address).await?;
 
-        let app = self.app;
-        let max_connections = self.max_connections;
-        let max_request_size = self.max_request_size;
-        let shutdown_timeout = self.shutdown_timeout;
+        if let Err(error) = listener.local_addr() {
+            let _ = &error;
+            // Placeholder for tracing...
+        }
 
-        listen(
-            acceptor,
-            address,
-            app,
-            max_connections,
-            max_request_size,
-            shutdown_timeout,
+        // Serve incoming connections until shutdown is requested.
+        serve(
+            Arc::clone(&self.state),
+            Arc::clone(&self.router),
+            self.acceptor()?,
```

Comment:
I like the idea of the `Server` struct being intact until the server exits. It's a best practice with async code. If having one additional reference to `state` or `router` ever becomes a problem, we can reconsider as this isn't a breaking change.